src/linker/Linker.Steps/SubStepsDispatcher.cs

```diff
@@ -0,0 +1,219 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+using Mono.Cecil;
+using Mono.Collections.Generic;
+
+namespace Mono.Linker.Steps
+{
+	//
+	// Generic steps dispatcher is intended to by used by custom linker step which
+	// consist of multiple steps. It simplifies their implementation as well as the
+	// way how to hook them into the pipeline of existing steps.
+	//
+	public abstract class SubStepsDispatcher : IStep, IEnumerable<ISubStep>
+	{
+		readonly List<ISubStep> substeps;
+
+		List<ISubStep> on_assemblies;
+		List<ISubStep> on_types;
+		List<ISubStep> on_fields;
+		List<ISubStep> on_methods;
+		List<ISubStep> on_properties;
+		List<ISubStep> on_events;
+
+		protected SubStepsDispatcher ()
+		{
+			substeps = new List<ISubStep> ();
+		}
+
+		protected SubStepsDispatcher (IEnumerable<ISubStep> subSteps)
+		{
+			substeps = new List<ISubStep> (subSteps);
+		}
+
+		public void Add (ISubStep substep)
+		{
+			substeps.Add (substep);
+		}
+
+		void IStep.Process (LinkContext context)
+		{
+			InitializeSubSteps (context);
+
+			BrowseAssemblies (context.GetAssemblies ());
+		}
+
+		static bool HasSubSteps (List<ISubStep> substeps) => substeps?.Count > 0;
+
+		void BrowseAssemblies (IEnumerable<AssemblyDefinition> assemblies)
+		{
+			foreach (var assembly in assemblies) {
+				CategorizeSubSteps (assembly);
+
+				if (HasSubSteps (on_assemblies))
+					DispatchAssembly (assembly);
+
+				if (!ShouldDispatchTypes ())
```

Comment:
If we think this is important to check up-front for perf reasons, then I would argue we should cache the "should" outside of the foreach. Similarly for the `HasSubSteps(on_assemblies)`.

Reply:
That is not possible as `CategorizeSubSteps` rebuilds the logic for every assembly

Reply:
hmm - OK. Not sure I like this approach, but it works.